src/workerd/api/r2-bucket.h

```diff
@@ -265,21 +294,135 @@ class R2Bucket: public jsg::Object {
   kj::Maybe<kj::String> adminBucket;
 
   friend class R2Admin;
+  friend class R2MultipartUpload;
+};
+
+enum class OptionalMetadata: uint16_t {
+  Http = static_cast<uint8_t>(R2ListRequest::IncludeField::HTTP),
+  Custom = static_cast<uint8_t>(R2ListRequest::IncludeField::CUSTOM),
 };
 
-#define EW_R2_PUBLIC_BETA_ISOLATE_TYPES \
-  api::R2Error, \
-  api::public_beta::R2Bucket, \
-  api::public_beta::R2Bucket::HeadResult, \
-  api::public_beta::R2Bucket::GetResult, \
-  api::public_beta::R2Bucket::Range, \
-  api::public_beta::R2Bucket::Conditional, \
-  api::public_beta::R2Bucket::GetOptions, \
-  api::public_beta::R2Bucket::PutOptions, \
-  api::public_beta::R2Bucket::Checksums, \
-  api::public_beta::R2Bucket::HttpMetadata, \
-  api::public_beta::R2Bucket::ListOptions, \
-  api::public_beta::R2Bucket::ListResult
-// The list of r2-bucket.h types that are added to worker.c++'s JSG_DECLARE_ISOLATE_TYPE
+template <typename T>
+concept HeadResultT = std::is_base_of_v<R2Bucket::HeadResult, T>;
+
+template <HeadResultT T, typename... Args>
+static jsg::Ref<T> parseObjectMetadata(R2HeadResponse::Reader responseReader,
```

Comment:
You want to be very careful about static methods in header because it's going to cause the compiler to inline it into every compilation unit that uses it. This is quite large so there's quite a bit of bloat happening here.

Would it work to revert the refactor and just expose a non-templated standalone function in the header that you call from r2-multipart.h instead?

Reply:
Done, thanks for the suggestion!